service-base/src/test/java/org/eclipse/hono/service/registration/AbstractRegistrationServiceTest.java

```diff
@@ -655,27 +592,23 @@ public void testUpdateSucceedWithMissingResourceVersion(final VertxTestContext c
     @Test
     public void testUpdateSucceedWithCorrectResourceVersion(final VertxTestContext ctx) {
 
-        final Promise<OperationResult<Id>> result = Promise.promise();
         final Checkpoint register = ctx.checkpoint(2);
         final JsonObject version = new JsonObject();
 
-        getDeviceManagementService().createDevice(TENANT, Optional.of(DEVICE), new Device(), NoopSpan.INSTANCE, result);
-        result.future()
-        .map(response -> {
-            assertEquals(HttpURLConnection.HTTP_CREATED, response.getStatus());
-            register.flag();
-            return response;
-        }).compose(rr -> {
-            final Promise<OperationResult<Id>> update = Promise.promise();
+        getDeviceManagementService().createDevice(TENANT, Optional.of(DEVICE), new Device(), NoopSpan.INSTANCE)
+                .map(response -> {
+                    assertEquals(HttpURLConnection.HTTP_CREATED, response.getStatus());
+                    register.flag();
+                    return response;
+                }).compose(rr -> {
             final String resourceVersion = rr.getResourceVersion().orElse(null);
             version.put("1", resourceVersion);
 
-            getDeviceManagementService().updateDevice(
-                    TENANT, DEVICE,
-                    new JsonObject().put("ext", new JsonObject().put("customKey", "customValue")).mapTo(Device.class),
-                    Optional.of(resourceVersion), NoopSpan.INSTANCE,
-                    update);
-            return update.future();
+            return getDeviceManagementService()
+                    .updateDevice(TENANT, DEVICE,
+                            new JsonObject().put("ext", new JsonObject().put("customKey", "customValue"))
+                                    .mapTo(Device.class),
+                            Optional.of(resourceVersion), NoopSpan.INSTANCE);
```

Comment:
Indentation in this method seems non-standard.

Something which was there in the code before already: Since `resourceVersion` can be null, `Optional.ofNullable(resourceVersion)` should be used, I guess.

Reply:
> Something which was there in the code before already: Since `resourceVersion` can be null, `Optional.ofNullable(resourceVersion)` should be used, I guess.

This has been already there and not added in this refactoring. Also in some places, I have seen scope for some optimisations and I don't want address this as part of this PR as this focuses mainly on refactoring the code to return future. IMHO a separate PR would be good. Also in the above test, _Optional.of(resourceVersion)_ is used to create an instance of `Optional` and for me it seems ok and I didn't get it why it should be `Optional.ofNullable(...)`

Reply:
> Also in the above test, _Optional.of(resourceVersion)_ is used to create an instance of `Optional` and for me it seems ok and I didn't get it why it should be `Optional.ofNullable(...)`

For `resourceVersion==null`, `Optional.of(resourceVersion)` will produce a NPE. But that change could also be put in another PR.